Cart Counter

supernova-at opened the pull request:
<!--
Before submitting this pull request, please make sure you have read our Contribution Guidelines and your PR meets our contribution standards:
https://github.com/magento/pwa-studio/blob/master/.github/CONTRIBUTING.md

Please fill out as much information as you can about your PR to help speed up the review process.
If your PR addresses an existing GitHub Issue, please refer to it in the title or Additional Information section to make the connection.

We may ask you for changes in your PR in order to meet the standards set in our Contribution Guidelines. PRs that do not comply with our guidelines may be closed at the maintainers' discretion.

Feel free to remove this section before creating this PR. Thank you for your contribution!
-->

## Description

The problem was that we were never requesting the cart's details on page load. Only when adding an item to the cart, updating an item in the cart, removing an item from the cart, or toggling the cart open did we request the cart's details.

Now we do.

This PR also updates components in the app's header:
  * `CartTrigger` calls `getCartDetails` when it gets mounted
  * `CartTrigger` is no longer a connected component
  * `Header/container.js` is the connected component, passing info down to `CartTrigger` via `header.js`
  * `CartTrigger` and `CartCounter` rewritten to be functional components
  * `CartTrigger` and `CartCounter` tests updated away from using `enzyme`

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here by replacing ISSUE_NUMBER with your actual issue number. -->
<!--- Using the above wording causes Github to automatically close the issue on merge. -->
Closes #1412.

## Verification Steps
<!-- Please describe in detail how a reviewer can verify your changes. -->
1. Load up Venia's homepage
1. If you have an item in your cart, see that the shopping cart shows the number of items correctly
1. If you do not have an item in your cart, add some and go back to step 1 😄 

## Screenshots / Screen Captures (if appropriate)

![Screen Shot 2019-08-13 at 4 30 29 PM](https://user-images.githubusercontent.com/13182778/62978784-b0093800-bde7-11e9-8c54-ba5596519fdb.png)

## Proposed Labels for Change Type/Package
<!--- What type of change level would you suggest for this PR? -->
<!--- Major, Minor, or Patch? -->
<!--- See https://pwastudio.io/technologies/versioning/ for help -->
- [ ] major (e.g x.0.0 - a breaking change)
- [x] minor (e.g 0.x.0 - a backwards compatible addition)
- [ ] patch (e.g 0.0.x - a bug fix)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have updated the documentation accordingly, if necessary.
- [x] I have added tests to cover my changes, if necessary.

sirugh commented on packages/venia-ui/lib/components/Header/cartTrigger.js:
```diff
@@ -1,75 +1,67 @@
-    }
-}
+    useEffect(() => {
+        if (getCartDetails) {
```
Is this a necessary conditional? Seems like the function should always be present.

supernova-at replied:
Yeah I wasn't really sure about this. It will always be present because I connected it that way, but I was thinking if someone else uses the `Header` component they may or may not connect it.

Marking this prop type as `.isRequired` is along these same lines.

I didn't mark it as `isRequired`, so I included this conditional.

supernova-at replied:
`Header` is [available as an export](https://github.com/magento/pwa-studio/blob/develop/packages/venia-ui/lib/components/Header/index.js#L1) from `Header/index.js`.

sirugh replied:
But `Header` passes the function now, so it _should_ always be defined.

I'd probably just mark it as `isRequired` and be done with.

supernova-at replied:
Resolved by https://github.com/magento/pwa-studio/pull/1556/commits/8bb2e7e3ff3feadfb28dbde4400f2a98d372536d.

sirugh commented on packages/venia-ui/lib/components/Header/cartTrigger.js:
```diff
@@ -1,75 +1,67 @@
+    }, [getCartDetails]);
+
+    const cartIcon = useMemo(() => {
+        return getCartIcon(cartDetails);
```
Minor improvement would be to pull off the `items_qty` from `cartDetails` as it seems like the only value the function cares about.

Beyond that we could also just hardcode the two possible icon states, one that has color and one that is filled and one that isn't.

supernova-at replied:
Resolved by https://github.com/magento/pwa-studio/pull/1556/commits/ced668878ed9b61219559142c28ecbfc2bff75d4.

sirugh commented on packages/venia-ui/lib/components/Header/container.js:
```diff
@@ -1,17 +1,20 @@
 
-import Header from './header';
 import { toggleSearch } from '../../actions/app';
+import { getCartDetails, toggleCart } from '../../actions/cart';
```
How much work would it be to write a new query/action that only requests the item count? I know the efficiency of our gql calls has been under the microscope lately so maybe a smaller query would help, especially for something that shows up on the header/initial load.

supernova-at replied:
I thought this was a great suggestion but unfortunately implementing it uncovered [a bug in our GraphQL](https://github.com/magento/graphql-ce/issues/830) where the item quantity is not being returned correctly.

sirugh replied:
😢

sirugh commented on packages/venia-ui/lib/components/Header/cartTrigger.js:
```diff
@@ -1,75 +1,67 @@
 
-        return <Icon src={ShoppingCartIcon} attrs={svgAttributes} />;
-    }
+const Trigger = props => {
```
Can we match this with the name of the file? There's already another `Trigger` or two in this repo :D

supernova-at replied:
Resolved by https://github.com/magento/pwa-studio/pull/1556/commits/ced668878ed9b61219559142c28ecbfc2bff75d4.

sirugh commented on packages/venia-ui/lib/components/Header/cartCounter.js:
```diff
@@ -1,23 +1,27 @@
-        ) : null;
+    const classes = mergeClasses(defaultClasses, props.classes);
+
+    // Don't show anything both when we don't have data to show, and when numItems is zero (0).
```
It appears that we used to _always_ show either 0 or the number of items. The conditional where we created this component passed the `item_qty` or `0`. This changes that behavior so I just want to call that out to ensure it was intentional.

supernova-at replied:
I don't think that has changed, `cartTrigger` had a [conditional in its render function](https://github.com/magento/pwa-studio/blob/develop/packages/venia-ui/lib/components/Header/cartCounter.js#L17-L19).

It shouldn't have ever displayed a zero.

sirugh replied:
Oh you're right! My bad :)

supernova-at commented:
I changed the recommended version to be `Minor` because according to our [Public API](https://github.com/magento/pwa-studio/blob/37cd2c18247eeb773677c960fa9620d2f98161fa/pwa-devdocs/src/technologies/versioning/index.md#public-api) policy, `Header/header.js` is part of the public API (it gets exported by [Header/index.js](https://github.com/magento/pwa-studio/blob/develop/packages/venia-ui/lib/components/Header/index.js#L1) ), and I added additional props to it.

dpatil-magento commented:
@supernova-at There is a unit test failure, can you please take a look?
packages/venia-ui/lib/components/Header/__tests__/cartTrigger.spec.js

supernova-at commented:
> @supernova-at There is a unit test failure, can you please take a look?
packages/venia-ui/lib/components/Header/tests/cartTrigger.spec.js

@dpatil-magento oops, forgot to update a snapshot. All set now, thanks!

dpatil-magento commented:
@supernova-at Fix looks good. Just curious to know why it takes ~10 seconds to show the count on page reload?

supernova-at commented:
> Just curious to know why it takes ~10 seconds to show the count on page reload?

It definitely should not!
